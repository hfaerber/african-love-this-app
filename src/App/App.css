*,
*:after,
*:before {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.App {
  text-align: center;
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-logo {
  height: 150vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-link {
  color: #61dafb;
}

.App-header {
margin: 5px;
}
/* @keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
} */

.div-giphy {
  width:13%;
  /* height:0; */
  padding-top: 4%;
  position:relative;
}

.giphy-embed {
  width: 100%;
  frameBorder: 0;
}
