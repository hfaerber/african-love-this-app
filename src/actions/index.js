export const updateUser = user => ({
  type: 'UPDATE_USER',
  user
});

export const updateCountries = countries => ({
  type: 'UPDATE_COUNTRIES',
  countries
});

export const toggleStudyStatus = name => ({
  type: 'TOGGLE_STUDY_STATUS',
  name
});

export const updateSelectedFilter = filter => ({
  type: 'UPDATE_SELECTED_FILTER',
  filter
});

export const updateSearchQuery = query => ({
  type: 'UPDATE_SEARCH_QUERY',
  query
});

export const updateError = errorMessage => ({
  type: 'UPDATE_ERROR',
  errorMessage
});

export const updateLoadingStatus = status => ({
  type: 'UPDATE_LOADING_STATUS',
  status
});
