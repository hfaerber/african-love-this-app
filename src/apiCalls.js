export const fetchCountryData = () => {
  return fetch('https://restcountries.eu/rest/v2/region/africa')
    .then(res => {
      if (!res.ok) {
        throw Error('something went wrong')
      }
      return res.json();
    })
}
