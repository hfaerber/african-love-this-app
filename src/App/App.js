import React, { Component } from 'react';
import './App.css';
import { fetchCountryData } from '../apiCalls.js';
import { cleanData } from '../util';
import WelcomeForm from '../WelcomeForm/WelcomeForm';
import StudyContainer from '../StudyContainer/StudyContainer';
import { Route } from 'react-router-dom';
import { updateCountries } from '../actions';
import { connect } from 'react-redux';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true
    }
  }

  componentDidMount() {
    fetchCountryData()
      .then(data => {
        let countries = cleanData(data);
        console.log(countries);
        this.setState({ isLoading: false });
        this.props.updateCountries(countries);
      }).catch(err => console.log(err))
  };

  render () {
    return (
      <main className="App">
        <header className="App-header">
          African <span role='img' aria-label='heart emoji' className='header-heart'>♥️</span> This App
        </header>
        {this.state.isLoading && <div className='div-giphy'><iframe src="https://giphy.com/embed/tkJsL5AIIsg7K" className="giphy-embed" allowFullScreen></iframe></div>}
        <Route exact path='/' component={WelcomeForm} />
        <Route path='/study' component={StudyContainer} />
      </main>
    );
  };
};

export const mapDispatchToProps = dispatch => ({
  updateCountries: countries => dispatch(updateCountries(countries))
});

export default connect(null, mapDispatchToProps)(App);

// <p><a href="https://giphy.com/gifs/oc-basins-watershedsriver-tkJsL5AIIsg7K">via GIPHY</a></p>
// Icons made by <a href="https://www.flaticon.com/authors/those-icons" title="Those Icons">Those Icons</a> from <a href="https://www.flaticon.com/" title="Flaticon"> www.flaticon.com</a>
// <div>Icons made by <a href="https://www.flaticon.com/authors/those-icons" title="Those Icons">Those Icons</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a></div>
// Icons made by <a href="https://www.flaticon.com/authors/srip" title="srip">srip</a> from <a href="https://www.flaticon.com/" title="Flaticon"> www.flaticon.com</a>
