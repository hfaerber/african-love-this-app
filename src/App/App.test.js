import React from 'react';
import { App, mapDispatchToProps } from './App';
import { shallow } from 'enzyme';
import { fetchCountryData } from '../apiCalls';
import { updateCountries, updateError, updateLoadingStatus } from '../actions';

jest.mock('../apiCalls');

describe('AppContainer', () => {
  describe('App component', () => {
    let wrapper, mockUpdateCountries, mockUpdateError, mockUpdateLoadingStatus;

    beforeEach(() => {
      mockUpdateCountries = jest.fn();
      mockUpdateError = jest.fn();
      mockUpdateLoadingStatus = jest.fn();
      fetchCountryData.mockImplementation(() => {
        return Promise.resolve( [{
          name: "Algeria", topLevelDomain: ['test'], alpha2Code: "DZ",
          callingCodes: [777], capital: "Algiers", subregion: "Northern Africa"
          }]
        );
      });
      wrapper = shallow(<App
        updateCountries={mockUpdateCountries}
        updateError={mockUpdateError}
        updateLoadingStatus={mockUpdateLoadingStatus}
      />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('should invoke fetchCountryData after component mounts', () => {
      expect(fetchCountryData).toHaveBeenCalled();
    });

    it('should invoke updateCountries and updateLoadingStatus props when fetchCountryData resolves', async () => {
      await fetchCountryData();
      expect(mockUpdateCountries).toHaveBeenCalled();
      expect(mockUpdateLoadingStatus).toHaveBeenCalledWith(false);
    });

    it('should invoke updateLoadingStatus and updateError if fetchCountryData rejects',
      async () => {
        fetchCountryData.mockImplementation(() => {
          return Promise.reject(Error('got an error'))
        });
        wrapper = await shallow(<App
          updateCountries={mockUpdateCountries}
          updateError={mockUpdateError}
          updateLoadingStatus={mockUpdateLoadingStatus}
        />);
        await wrapper.instance().forceUpdate();
        expect(mockUpdateError).toHaveBeenCalledWith('Error loading study cards. Please refresh to try again.');
        expect(mockUpdateLoadingStatus).toHaveBeenCalledWith(false);
    });

  });

  describe('mapDispatchToProps', () => {
    let mockDispatch, mappedProps;
    beforeEach(()=> {
      mockDispatch = jest.fn();
      mappedProps = mapDispatchToProps(mockDispatch);
    });

    it('should call dispatch with an updateCountries action when updateCountries is called', () => {
      const mockCountries = [
        {name: "Algeria", capital: "Algiers", subregion: "Northern Africa", needsStudied: true},
        {name: "Benin", capital: "Porto-Novo", subregion: "Western Africa", needsStudied: true},
      ];
      const actionToDispatch = updateCountries(mockCountries);
      mappedProps.updateCountries(mockCountries);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an updateError action when updateError is called', () => {
      const errorMessage = 'error';
      const actionToDispatch = updateError(errorMessage);
      mappedProps.updateError(errorMessage);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an updateLoadingStatus action when updateLoadingStatus is called', () => {
      const status = false;
      const actionToDispatch = updateLoadingStatus(status);
      mappedProps.updateLoadingStatus(status);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  })
})
