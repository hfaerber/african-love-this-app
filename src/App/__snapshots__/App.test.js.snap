// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer App component should match snapshot 1`] = `
<main
  className="App"
>
  <header
    className="App-header"
  >
    African 
    <span
      aria-label="heart emoji"
      className="header-heart"
      role="img"
    >
      ♥️
    </span>
     This App
  </header>
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(WelcomeForm)",
          "type": [Function],
        }
      }
      exact={true}
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(StudyContainer)",
          "type": [Function],
        }
      }
      path="/study"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</main>
`;
